### PR TITLE
Add kineto and autograd (emit_nvtx) profiling

### PR DESCRIPTION
**Pull request by jithunnair-amd**
Enabled:
1. Kineto profiler using the `--kineto` flag
2. Autograd profiler (emit_nvtx) using the `--autograd_profiler` flag
 
### Kineto profiling results:

On MI60, the following slowdown was observed with the following profiler scheduler:
```
        profiler_schedule = schedule(
            skip_first = 0,
            wait = 0,
            warmup = 0,
            active = 20,
            repeat = 1,
        )
```

Command: 

**Without kineto**: `python3 micro_benchmarking_pytorch.py --network resnet50 --iterations 20`
```
--------------------SUMMARY--------------------------
Microbenchmark for network : resnet50
Num devices: 1
Dtype: FP32
Mini batch size [img] : 64
Time per mini-batch : 0.2259697437286377
Throughput [img/sec] : 283.2237579419316
```

**With kineto**: `python3 micro_benchmarking_pytorch.py --network resnet50 --iterations 20 --kineto`
```
--------------------SUMMARY--------------------------
Microbenchmark for network : resnet50
Num devices: 1
Dtype: FP32
Mini batch size [img] : 64
Time per mini-batch : 0.9641143083572388
Throughput [img/sec] : 66.38217008629408
```

That's a 77% slowdown! :/

Here's the kineto profiling results with above run:
```
/opt/conda/lib/python3.7/site-packages/torch/profiler/profiler.py:224: UserWarning: Profiler won't be using warmup, this can skew profiler results
  warn("Profiler won't be using warmup, this can skew profiler results")
----------- Trace Ready -----------
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                          ProfilerStep*         0.01%     890.000us        57.96%        5.885s     294.245ms       1.206ms         0.01%        5.885s     294.233ms            20
autograd::engine::evaluate_function: ConvolutionBack...         0.28%      28.674ms        10.06%        1.021s     963.520us      15.657ms         0.11%        2.687s       2.535ms          1060
                                   ConvolutionBackward0         1.90%     192.479ms         9.66%     981.144ms     925.608us      13.399ms         0.09%        2.560s       2.415ms          1060
                             aten::convolution_backward         6.70%     680.325ms         7.77%     788.665ms     744.024us        2.516s        17.81%        2.547s       2.403ms          1060
                                           aten::conv2d         1.43%     145.602ms         7.13%     723.917ms     682.941us      13.264ms         0.09%        1.008s     951.158us          1060
                                      aten::convolution         0.25%      25.126ms         5.70%     578.315ms     545.580us      13.384ms         0.09%     994.964ms     938.645us          1060
                                     aten::_convolution         3.99%     404.852ms         5.45%     553.189ms     521.876us      13.379ms         0.09%     981.580ms     926.019us          1060
                               aten::miopen_convolution         1.05%     106.706ms         1.46%     148.337ms     139.941us     955.858ms         6.77%     968.201ms     913.397us          1060
autograd::engine::evaluate_function: MiopenBatchNorm...         1.44%     146.709ms        13.26%        1.347s       1.271ms      30.055ms         0.21%     494.231ms     466.256us          1060
                               MiopenBatchNormBackward0         2.63%     267.078ms        11.82%        1.200s       1.132ms      13.991ms         0.10%     464.176ms     437.902us          1060
                       aten::miopen_batch_norm_backward         1.50%     152.357ms         9.19%     932.983ms     880.173us     424.869ms         3.01%     450.185ms     424.703us          1060
                              MIOpenBatchNormBwdSpatial         0.00%       0.000us         0.00%       0.000us       0.000us     384.780ms         2.72%     384.780ms     363.000us          1060
                                Optimizer.step#SGD.step         0.60%      60.599ms         3.70%     375.904ms      18.795ms     146.429ms         1.04%     367.759ms      18.388ms            20
                                       aten::batch_norm         0.22%      22.703ms         5.04%     511.687ms     482.724us      13.089ms         0.09%     330.501ms     311.793us          1060
                                             aten::add_         4.27%     433.304ms         4.27%     433.304ms      39.249us     320.844ms         2.27%     320.844ms      29.062us         11040
                           aten::_batch_norm_impl_index         0.36%      36.319ms         4.82%     488.984ms     461.306us      20.287ms         0.14%     317.412ms     299.445us          1060
                                           iteration 19         2.06%     209.150ms         3.12%     317.084ms     317.084ms     208.361ms         1.47%     317.069ms     317.069ms             1
                                   miopenSp3AsmConvRxSU         0.00%       0.000us         0.00%       0.000us       0.000us     311.302ms         2.20%     311.302ms     915.594us           340
                                            iteration 4         1.84%     186.726ms         2.99%     303.757ms     303.757ms     184.197ms         1.30%     303.753ms     303.753ms             1
                                           iteration 16         1.93%     196.301ms         2.99%     303.520ms     303.520ms     195.461ms         1.38%     303.511ms     303.511ms             1
                                            iteration 2         1.86%     188.599ms         2.98%     302.671ms     302.671ms     186.216ms         1.32%     302.660ms     302.660ms             1
                                            iteration 0         1.84%     186.743ms         2.97%     301.378ms     301.378ms     184.588ms         1.31%     301.354ms     301.354ms             1
                                            iteration 3         1.82%     185.227ms         2.91%     295.524ms     295.524ms     182.928ms         1.29%     295.512ms     295.512ms             1
                                            iteration 9         1.84%     186.951ms         2.90%     294.310ms     294.310ms     186.172ms         1.32%     294.316ms     294.316ms             1
                                            iteration 1         1.81%     184.212ms         2.89%     293.173ms     293.173ms     183.073ms         1.30%     293.158ms     293.158ms             1
                                            iteration 5         1.83%     185.997ms         2.88%     292.729ms     292.729ms     183.905ms         1.30%     292.713ms     292.713ms             1
                                           iteration 18         1.82%     184.439ms         2.87%     291.621ms     291.621ms     183.619ms         1.30%     291.613ms     291.613ms             1
                                           iteration 17         1.81%     183.957ms         2.87%     291.077ms     291.077ms     182.946ms         1.29%     291.066ms     291.066ms             1
                                aten::miopen_batch_norm         2.20%     223.663ms         4.13%     419.345ms     395.608us     266.085ms         1.88%     290.944ms     274.475us          1060
                                            iteration 6         1.81%     183.808ms         2.85%     289.038ms     289.038ms     182.919ms         1.29%     289.030ms     289.030ms             1
                                           iteration 11         1.81%     183.864ms         2.85%     288.956ms     288.956ms     183.073ms         1.30%     288.943ms     288.943ms             1
                                            iteration 7         1.81%     183.414ms         2.85%     288.946ms     288.946ms     182.504ms         1.29%     288.935ms     288.935ms             1
                                            iteration 8         1.81%     183.413ms         2.84%     288.831ms     288.831ms     182.520ms         1.29%     288.820ms     288.820ms             1
                                           iteration 13         1.81%     183.286ms         2.84%     288.320ms     288.320ms     182.594ms         1.29%     288.307ms     288.307ms             1
                                           iteration 15         1.81%     183.702ms         2.84%     288.229ms     288.229ms     182.834ms         1.29%     288.235ms     288.235ms             1
                                           iteration 12         1.81%     183.554ms         2.84%     287.938ms     287.938ms     182.836ms         1.29%     287.929ms     287.929ms             1
                                           iteration 10         1.81%     183.942ms         2.84%     287.849ms     287.849ms     183.089ms         1.30%     287.840ms     287.840ms             1
                                           iteration 14         1.81%     183.311ms         2.83%     287.648ms     287.648ms     182.556ms         1.29%     287.638ms     287.638ms             1
void at::native::modern::elementwise_kernel<at::nati...         0.00%       0.000us         0.00%       0.000us       0.000us     263.440ms         1.86%     263.440ms      25.577us         10300
                                  MIOpenCvBwdWrW_8x8map         0.00%       0.000us         0.00%       0.000us       0.000us     237.394ms         1.68%     237.394ms       2.374ms           100
igemm_v4r1_dynamic_wrw_128x128x16_8x8_4x4x4x4x4x4_16...         0.00%       0.000us         0.00%       0.000us       0.000us     
.....

-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 10.153s
Self CUDA time total: 14.129s
``` 

### Autograd profiler results:

On MI60:
Command: 
**Without autograd profiler**: `python3 micro_benchmarking_pytorch.py --network resnet50 --iterations 20`
```
INFO: running forward and backward for warmup.
INFO: running the benchmark..
OK: finished running benchmark..
--------------------SUMMARY--------------------------
Microbenchmark for network : resnet50
Num devices: 1
Dtype: FP32
Mini batch size [img] : 64
Time per mini-batch : 0.21925389766693115
Throughput [img/sec] : 291.8990297596555
```

**With autograd profiler**: `python3 micro_benchmarking_pytorch.py --network resnet50 --iterations 20 --autograd_profiler`
```
INFO: running forward and backward for warmup.
INFO: running the benchmark..
OK: finished running benchmark..
--------------------SUMMARY--------------------------
Microbenchmark for network : resnet50
Num devices: 1
Dtype: FP32
Mini batch size [img] : 64
Time per mini-batch : 0.2328105092048645
Throughput [img/sec] : 274.90167956156307
```

That's a 6% slowdown just with autograd_profiler emit_nvtx turned on.

**With autograd profiler AND rpd_tracer**: `runTracer.sh python micro_benchmarking_pytorch.py --network resnet50 --iterations 20 --autograd_profiler`
```
Creating empty rpd: trace.rpd
rpd_tracer, because
rpd_tracer, because
INFO: running forward and backward for warmup.
INFO: running the benchmark..
OK: finished running benchmark..
--------------------SUMMARY--------------------------
Microbenchmark for network : resnet50
Num devices: 1
Dtype: FP32
Mini batch size [img] : 64
Time per mini-batch : 0.2582177400588989
Throughput [img/sec] : 247.85283917906546
rocpd_string: 0
rocpd_op: 0
rocpd_api_ops: 0
rocpd_kernelapi: 1
rocpd_copyapi: 1
rocpd_api: 0
rpd_tracer: finalized in 181.501499 ms
```

That's a 10% further slowdown (and a 15% slowdown overall) due to turning on rpd_tracer.

### CODE REVIEW

**Review by jeffdaily — APPROVED** [review]
LGTM

**Comment by jithunnair-amd** [comment]
Also added autograd emit_nvtx profiling using the `--autograd_profiler` flag (mutually exclusive and lower priority to `--kineto`). Ran with rpd_tracer as follows:
`runTracer.sh python micro_benchmarking_pytorch.py --network resnet50 --autograd_profiler`

**Review by jeffdaily — APPROVED** [review]
Approved again.

**Comment by jeffdaily** [comment]
Any confirmation whether the performance running without profiling was affected in any way by the PR changes?

**Comment by jithunnair-amd** [comment]
> Any confirmation whether the performance running without profiling was affected in any way by the PR changes?

@jeffdaily @mwootton Yes, I verified by running 1. without the `autograd.profiler.emit_nvtx` context manager and 2. with the context manager but with `enabled=False`, and I got the same performance for both cases.

File: micro_benchmarking_pytorch.py [hunk 1]
```diff
@@ -215,13 +215,39 @@ def run_benchmarking(local_rank, ngpus, net, batch_size, iterations, prof_step,
...
+            torch.cuda.synchronize()
+            print(prof.key_averages().table(sort_by="cuda_time_total"))
+    else:
+        tm = time.time()
+        with torch.autograd.profiler.emit_nvtx(enabled=args.autograd_profiler):
+            for i in range(iterations):
+                if i == flops_prof_step:
+                    forwardbackward(inp, optimizer, network, target, amp_opt_level, i)
+                else:
+                    forwardbackward(inp, optimizer, network, target, amp_opt_level)
```
Comment (jithunnair-amd) [inline micro_benchmarking_pytorch.py:250]: @jeffdaily @mwootton We missed a bug here: I accidentally removed the `torch.cuda.synchronize()` call in the else (autograd_profiler) clause, causing it to undercount the time (eg. https://ontrack-internal.amd.com/browse/SWDEV-341071?focusedCommentId=9009931&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-9009931). I fixed the bug here: https://github.com/ROCmSoftwarePlatform/pytorch-micro-benchmarking/commit/7542f973c737dccd5e46d0a9cce557971dfcc2f3

**Pull request merged**